### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by satazor**
The goal is to enable this plugin in our own `eslint-config-uphold`. This way, we will have validation for backend `require` statements order similar to the frontend. Please note, however, we will have to change `lodash` order to be in the right place. Currently, it's always in the first position but it's an exception.

------
This plugin was bootstrapped using the official plugin generator. The generator sets up the folder structure, package.json, README, tests (with mocha), etc. I kept these mostly as-is for consistency with the overall ESLint community.

### CODE REVIEW

**Comment by satazor** [comment]
@nunofgs @franciscocardoso what's the expected minimum Node.js version? At the moment this plugin works on `>= v14`. Is that fine?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        node: ['14', '16']
+    name: Node ${{ matrix.node }}
+
+    steps:
+      - uses: actions/checkout@v1
+      - uses: actions/setup-node@v1
```
Comment (diogotorres97) [inline .github/workflows/ci.yml:15]:
```suggestion
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
```

Comment (satazor) [reply .github/workflows/ci.yml]: Fixed!

File: tests/sort-requires.js [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+        { message: "Expected require declaration of 'A' to be placed before 'b'.", type: 'VariableDeclaration' }
+      ],
+      options: [{ autofixDeclarationSortOrder: true }],
+      output: `
+        const { c, e } = require('foo');
+        const { d, f } = require('foo');
+        const A = require('foo');
+        const { ...Z } = require('foo');
+        const a = require('foo');
+        const { b } = require('foo');
```
Comment (Americas) [inline tests/sort-requires.js:97]: I would argue that this require should be the first one, it is using the multiple syntax even if just declaring one.

File: tests/sort-requires.js [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+      options: [{ autofixDeclarationSortOrder: true, useOldSingleMemberSyntax: true }],
+      output: `
+        const { a } = require('foo');
+        const c = require('foo');
+      `
+    },
+    {
+      code: `
+        const c = require('foo');
+        const { ...Z } = require('foo');
```
Comment (Americas) [inline tests/sort-requires.js:212]: Can't we auto fix this to `const Z = require('foo');`? :joy:

File: .eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (satazor) [reply .eslintrc.js:1]: No specific reason, I'm just more used to JSON and I find it more readable. Want me to change it?

File: tests/sort-imports.js [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+        { message: "Member 'C' of the import declaration should be sorted alphabetically.", type: 'ImportSpecifier' },
+        { message: "Member 'D' of the import declaration should be sorted alphabetically.", type: 'ImportSpecifier' },
+        { message: "Member 'E' of the import declaration should be sorted alphabetically.", type: 'ImportSpecifier' }
+      ]
+    },
+    // Member sort order - `ignoreCase` option.
+    {
+      code: `
+        import { b, A } from 'foo';
+        import { d, C } from 'foo';
```
[the remark this answers is not in the document]
Comment (satazor) [reply tests/sort-imports.js:463]: Yep, changing

File: tests/sort-requires.js (hunk 2 above)
Comment (satazor) [reply tests/sort-requires.js:97]: That's the case when `useOldSingleMemberSyntax` option is on. We will turn it on in our `eslint-config-uphold`.

File: tests/sort-requires.js (hunk 3 above)
Comment (satazor) [reply tests/sort-requires.js:212]: Ahaha I guess it's out of the scope of this project :D

File: lib/rules/sort-requires.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+const isTopLevel = node => node.parent.type === 'Program';
+
+const isStaticRequire = node =>
+  node.type === 'CallExpression'
+    ? node.callee.type === 'Identifier' && node.callee.name === 'require' && node.arguments.length === 1
+    : false;
+
+const isRequire = node =>
+  node.type === 'VariableDeclaration' ? node.declarations[0].init?.callee?.name === 'require' : false;
```
[the remark this answers is not in the document]
Comment (satazor) [reply lib/rules/sort-requires.js:11]: Changing.

File: lib/rules/sort-requires.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
+const buildSortRule = require('../utils/sort-rule-builder');
+
+const isTopLevel = node => node.parent.type === 'Program';
+
+const isStaticRequire = node =>
+  node.type === 'CallExpression'
+    ? node.callee.type === 'Identifier' && node.callee.name === 'require' && node.arguments.length === 1
+    : false;
```
[the remark this answers is not in the document]
Comment (satazor) [reply lib/rules/sort-requires.js:8]: Changing.

**Comment by satazor** [comment]
Thanks for the review @Americas 🙏. Just one unsolved comment.

File: .eslintrc.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Americas) [reply .eslintrc.js:1]: I was wondering why, it is fine as it is, especially if it follows the community practice :smile:

File: lib/rules/sort-requires.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  node.type === 'CallExpression' &&
+  node.callee.type === 'Identifier' &&
+  node.callee.name === 'require' &&
+  node.arguments.length === 1;
+
+const isRequire = node => node.type === 'VariableDeclaration' && node.declarations[0].init?.callee?.name === 'require';
+
+const hasVariableDeclarations = node => !!node.declarations;
+
+const hasObjectPattern = node => node.declarations?.[0].id?.type === 'ObjectPattern';
```
Comment (acostalima) [inline lib/rules/sort-requires.js:15]:
```suggestion
const hasObjectPattern = node => node.declarations?.[0]?.id?.type === 'ObjectPattern';
```

I'm not sure whether this change makes sense though. 😄

File: lib/rules/sort-requires.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+const hasMultipleProperties = node => node.declarations?.[0].id?.properties.length > 1;
```
Comment (acostalima) [inline lib/rules/sort-requires.js:17]:
```suggestion
const hasMultipleProperties = node => node.declarations?.[0]?.id?.properties.length > 1;
```

I'm not sure whether this change makes sense though. 😄

**Review by acostalima — COMMENTED** [review]
Well done! 👏🏻 ❤️

File: .eslintrc.js [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+  extends: ['uphold'],
```
Comment (acostalima) [inline .eslintrc.js:2]: Don't we need to add `eslint-config-uphold` to the project?

Comment (satazor) [reply .eslintrc.js:2]: Technically, it `uphold-scripts` should have a peer dependency on our eslint config, and eslint as well. However, it’s currently a normal dependency and we are relying on the fact that the dependencies will be deduped and installed at the root. This is mostly fine.

File: lib/rules/sort-requires.js (hunk 9 above)
Comment (satazor) [reply lib/rules/sort-requires.js:17]: That’s not needed because declarations is never an empty array.

File: lib/rules/sort-requires.js (hunk 8 above)
Comment (satazor) [reply lib/rules/sort-requires.js:15]: Same

File: README.md [hunk 11]
```diff
@@ -1 +1,101 @@
...
+Whether to autofix potentially unsafe scenarios automatically when the `--fix` flag is used when calling `eslint`.
+
+The current scenarios considered unsafe are:
+
+- Sorting import / require declarations because they can have side-effects, therefore the order in which they are executed might matter. That's the reason why the built-in ESLint [`sort-imports`](https://eslint.org/docs/rules/sort-imports) rule does autofix their order.
+- Sorting dynamic keys with potential side-effects, e.g.: `const { [foo()]: bar } = require('bar')`. In this scenario, the order in which keys are declared might matter.
+
+Enable this option at your own discretion.
+
+#### `useLocalNames`
```
[the remark this answers is not in the document]
Comment (satazor) [reply README.md:78]: That’s a good question. I followed the name on the AST, which is local.name. By having name in the option makes one thing of named exported as well. Not sure.

[the remark this answers is not in the document]
Comment (threequartersjohn) [reply README.md:78]: Some notes here:
- Personally, `Local` makes me think of a nested scope
- `localName` [already exists](https://developer.mozilla.org/en-US/docs/Web/API/Element/localName) in what could be considered a similar context
- A google search for `javascript alias` gives us what we're looking for here, but `javascript local name` does not 😄

[the remark this answers is not in the document]
Comment (satazor) [reply README.md:78]: Alright I will change to `useAliases`

[the remark this answers is not in the document]
Comment (satazor) [reply README.md]: Changed.

**Pull request merged**